### PR TITLE
Merge dataset into dev

### PR DESCRIPTION
**Pull request by USSiamaboat**

### CODE REVIEW

**Comment by USSiamaboat** [comment]
Force pushed to undo bad merge

**Comment by USSiamaboat** [comment]
New `dataset.py` breaks the code in `train.py`, which relies on the old `dataset.py`, but the code in `train.py` doesn't run anyways

File: src/model/preprocess.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    return False
+
+
+# Filters samples to return only code
+def keep_only_content(sample: dict) -> dict:
+    # TODO: Is there a way to remove other keys instead of forcing []?
+    for key in sample.keys():
+        if key == "content":
+            continue
+        sample[key] = []
```
Comment (danaiamirali) [inline src/model/preprocess.py]: Could use `pop` to drop the keys we don't want. Also, should make the if condition check for `key != "content"` instead of `key == "content"` since `!=` is the common case
```suggestion
    for key in sample.keys():
        if key != "content":
            sample[key].pop()
```

Comment (danaiamirali) [reply src/model/preprocess.py]: In general though I would question whether this function is really useful. Not sure if the work of iterating over the dict keys for each element is worth it given that this is a streaming dataset. Whether this is useful depends on `.map` implementation

File: src/model/dataset.py [hunk 2]
```diff
@@ -1,109 +1,133 @@
...
-class SafeIterableDataset(torch.utils.data.IterableDataset):
-    """Wrapper to account for download errors so training doesn't stop due to error pulling data from HF."""
-    def __init__(self, dataset):
-        self.dataset = dataset
-
-    def __iter__(self):
-        iterator = iter(self.dataset)
         while True:
+            # Respect max_size
+            if count == self.max_size:
```
Comment (danaiamirali) [inline src/model/dataset.py]: Not sure why using a count is being used over `ds.take(max_size)` in the constructor.

Achieves the same thing but you should always delegate the work to be done by HuggingFace. There may be optimizations in `.take` taking advantage of the fact that this dataset is a streaming dataset that using a counter in a wrapper around the entire dataset might not utilze.

File: src/model/dataset.py [hunk 3]
```diff
@@ -1,109 +1,133 @@
...
+                break
+
+
+# SAMPLE USAGE
+if __name__ == "__main__":
+    tokenizer = AutoTokenizer.from_pretrained("./tokenizer_10M")
+
+    ds = ChunkedDataset(
+        train_split=True,      # Use training split
+        max_size=1_000_000,    # Provide up to 1 million samples (not files)
```
Comment (danaiamirali) [inline src/model/dataset.py:116]: samples vs. files is a good distinction

File: src/model/dataset.py (hunk 2 above)
Comment (danaiamirali) [reply src/model/dataset.py]: Especially since it seems `count` here is unrelated to `count` in ChunkingDataset. The usage of `count` makes sense in ChunkingDataset given the distinction between samples and files, but there is no such distinction in CleanDataset.

File: src/model/preprocess.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        "sklearn",
+        "tensorflow",
+        "torch",
+        "scipy",
+    ]
+    for library in libraries:
+        a = "import " + library
+        b = "from " + library + " import"
+        if a in content or b in content:
+            return True
```
Comment (danaiamirali) [inline src/model/preprocess.py]: This is unnecessarily expensive. Just looking for whether `library in content` would suffice

```suggestion
    for library in libraries:
        if library in content:
            return True
```

**Comment by danaiamirali** [comment]
Also `tokenizer_10m/` should not be getting pushed to git. Maybe a tarball of that folder could, but the actual folder should not be getting pushed

File: src/model/preprocess.py (hunk 1 above)
Comment (USSiamaboat) [reply src/model/preprocess.py]: Implemented better version in https://github.com/MichiganDataScienceTeam/F24-mini-copilot/commit/48b7f4d66708990cf67b64f1afa218721a3dc67d

File: src/model/dataset.py (hunk 2 above)
Comment (USSiamaboat) [reply src/model/dataset.py]: Implemented in https://github.com/MichiganDataScienceTeam/F24-mini-copilot/commit/a9bb0af1b090754c624e6850bbf2989690568c2d

File: src/model/preprocess.py (hunk 4 above)
Comment (USSiamaboat) [reply src/model/preprocess.py]: Implemented with regex in https://github.com/MichiganDataScienceTeam/F24-mini-copilot/commit/48b7f4d66708990cf67b64f1afa218721a3dc67d

**Comment by USSiamaboat** [comment]
> Also `tokenizer_10m/` should not be getting pushed to git. Maybe a tarball of that folder could, but the actual folder should not be getting pushed

Implemented the tar thing for now. Where would the files go if we don't put it in the github?

**Comment by danaiamirali** [comment]
Typically would use AWS S3 for something like this but we can use Drive for our project. Gave you access to a `data` folder in the Mini Copilot drive folder
